p2p/src/simulated/network.rs

```diff
@@ -517,22 +525,77 @@ impl Peer {
         socket: SocketAddr,
         max_size: usize,
     ) -> Self {
+        // Create inbound handler
+        let (inbound, mut outbound) = mpsc::unbounded();
+        let (control, mut control_receiver) = mpsc::unbounded();
+
+        // Spawn router
+        runtime.spawn("router", async move {
+            let mut mailboxes = HashMap::new();
+            loop {
+                select! {
+                    control = control_receiver.next() => {
+                        // If control is closed, exit
+                        let (channel, result): (Channel, oneshot::Sender<Result<UnboundedReceiver<Message>, Error>>) = match control {
+                            Some(control) => control,
+                            None => break,
+                        };
+
+                        // Check if channel is registered
+                        if mailboxes.contains_key(&channel) {
+                            result.send(Err(Error::ChannelAlreadyRegistered(channel))).unwrap();
+                            continue;
+                        }
+
+                        // Register channel
+                        let (sender, receiver) = mpsc::unbounded();
+                        mailboxes.insert(channel, sender);
+                        result.send(Ok(receiver)).unwrap();
+                    },
+                    message = outbound.next() => {
+                        // If outbound is closed, exit
+                        let message = match message {
+                            Some(message) => message,
+                            None => break,
+                        };
+
+                        // Send message to mailbox
+                        let (channel, message) = message;
```

Comment:
Can probably do away with this (and just pull from original match) -> can also do an `let Some() = message else{}`